In [70]:
import pandas as pd
from pathlib import Path
from am_notebooks import nb_show_table

from source.utils.general import SANPI_HOME

subtotals = pd.concat([pd.read_csv(csv, index_col='part').astype('int').sort_index().T 
             for csv in SANPI_HOME.joinpath('info/final_subtotals').glob('*.csv')]).T
nb_show_table(subtotals.convert_dtypes())

    


|           |   `POSmirror_hits` |   `mirrorNEQ_hits` |   `not-RBdirect_hits` |   `superNEQ_hits` |   `NEGmirror_hits` |   `RBdirect_hits` |
|:----------|-------------------:|-------------------:|----------------------:|------------------:|-------------------:|------------------:|
| **Apw**   |             17,294 |              3,805 |             1,878,981 |            86,805 |              5,320 |            90,499 |
| **Nyt1**  |             26,803 |              5,474 |             1,937,910 |            89,336 |             10,796 |            98,361 |
| **Nyt2**  |             41,008 |              8,679 |             2,746,650 |           126,486 |             14,761 |           136,138 |
| **Pcc00** |             43,742 |              9,078 |             2,100,614 |            96,962 |              8,541 |            95,376 |
| **Pcc01** |             43,095 |              9,151 |             2,068,435 |            95,634 |              8,498 |            94,471 |
| **Pcc02** 


|           |   `POSmirror_hits` |   `mirrorNEQ_hits` |   `not-RBdirect_hits` |   `superNEQ_hits` |   `NEGmirror_hits` |   `RBdirect_hits` |
|:----------|-------------------:|-------------------:|----------------------:|------------------:|-------------------:|------------------:|
| **Apw**   |             17,294 |              3,805 |             1,878,981 |            86,805 |              5,320 |            90,499 |
| **Nyt1**  |             26,803 |              5,474 |             1,937,910 |            89,336 |             10,796 |            98,361 |
| **Nyt2**  |             41,008 |              8,679 |             2,746,650 |           126,486 |             14,761 |           136,138 |
| **Pcc00** |             43,742 |              9,078 |             2,100,614 |            96,962 |              8,541 |            95,376 |
| **Pcc01** |             43,095 |              9,151 |             2,068,435 |            95,634 |              8,498 |            94,471 |
| **Pcc02** |             42,597 |              9,057 |             2,032,504 |            93,915 |              8,585 |            92,869 |
| **Pcc03** |             43,303 |              8,903 |             2,073,494 |            95,492 |              8,739 |            94,950 |
| **Pcc04** |             43,191 |              9,013 |             2,071,538 |            95,210 |              8,679 |            94,190 |
| **Pcc05** |             43,447 |              9,064 |             2,082,218 |            96,276 |              8,813 |            95,371 |
| **Pcc06** |             45,422 |              9,637 |             2,158,603 |            99,462 |              9,061 |            98,355 |
| **Pcc07** |             43,161 |              9,008 |             2,051,873 |            94,592 |              8,673 |            94,404 |
| **Pcc08** |             44,889 |              9,503 |             2,140,164 |            98,722 |              8,958 |            96,890 |
| **Pcc09** |             43,195 |              8,959 |             2,063,787 |            95,298 |              8,569 |            94,183 |
| **Pcc10** |             42,785 |              9,085 |             2,050,397 |            94,563 |              8,597 |            94,632 |
| **Pcc11** |             42,975 |              8,945 |             2,067,797 |            95,398 |              8,552 |            94,501 |
| **Pcc12** |             43,871 |              9,220 |             2,106,105 |            97,295 |              8,921 |            97,305 |
| **Pcc13** |             42,490 |              9,002 |             2,025,964 |            93,235 |              8,562 |            92,541 |
| **Pcc14** |             42,910 |              8,941 |             2,030,054 |            93,298 |              8,640 |            93,545 |
| **Pcc15** |             42,575 |              9,025 |             2,027,709 |            93,609 |              8,447 |            92,783 |
| **Pcc16** |             44,291 |              9,245 |             2,093,696 |            96,596 |              8,685 |            95,411 |
| **Pcc17** |             42,742 |              8,930 |             2,048,259 |            94,653 |              8,410 |            93,898 |
| **Pcc18** |             43,988 |              9,194 |             2,117,053 |            97,308 |              8,764 |            96,685 |
| **Pcc19** |             44,826 |              9,406 |             2,139,234 |            98,870 |              9,181 |            97,836 |
| **Pcc20** |             44,034 |              9,329 |             2,102,322 |            96,686 |              8,818 |            96,346 |
| **Pcc21** |             42,806 |              8,971 |             2,038,043 |            93,481 |              8,480 |            93,045 |
| **Pcc22** |             43,859 |              9,223 |             2,080,162 |            95,861 |              8,840 |            95,906 |
| **Pcc23** |             44,082 |              9,228 |             2,094,183 |            96,892 |              8,745 |            95,624 |
| **Pcc24** |             43,867 |              9,209 |             2,077,021 |            96,231 |              8,714 |            95,410 |
| **Pcc25** |             42,977 |              9,045 |             2,039,960 |            94,028 |              8,582 |            93,534 |
| **Pcc26** |             42,132 |              8,838 |             2,014,969 |            92,834 |              8,619 |            92,787 |
| **Pcc27** |             42,081 |              8,824 |             2,013,996 |            92,897 |              8,422 |            92,345 |
| **Pcc28** |             43,311 |              9,187 |             2,061,840 |            95,783 |              8,675 |            94,606 |
| **Pcc29** |             42,438 |              8,990 |             2,020,064 |            93,800 |              8,500 |            92,838 |
| **PccTe** |              1,371 |                274 |                66,019 |             3,108 |                291 |             2,970 |
| **PccVa** |              1,340 |                293 |                66,074 |             3,065 |                297 |             3,076 |



In [71]:
nb_show_table(subtotals.describe().T.round(1).convert_dtypes().sort_values('mean'), n_dec=1)


|                       |   `count` |      `mean` |     `std` |   `min` |       `25%` |     `50%` |       `75%` |     `max` |
|:----------------------|----------:|------------:|----------:|--------:|------------:|----------:|------------:|----------:|
| **mirrorNEQ_hits**    |        35 |     8,335.3 |   2,284.2 |     274 |     8,935.5 |     9,025 |     9,201.5 |     9,637 |
| **NEGmirror_hits**    |        35 |     8,335.3 |   2,365.4 |     291 |     8,520.5 |     8,640 |     8,788.5 |    14,761 |
| **POSmirror_hits**    |        35 |    39,682.8 |  10,892.7 |   1,340 |    42,532.5 |    43,095 |    43,863.0 |    45,422 |
| **superNEQ_hits**     |        35 |    90,676.6 |  22,644.0 |   3,065 |    93,545.0 |    95,298 |    96,641.0 |   126,486 |
| **RBdirect_hits**     |        35 |    90,676.6 |  23,052.2 |   2,970 |    92,957.0 |    94,501 |    95,765.0 |   136,138 |
| **not-RBdirect_hits** |        35 | 1,965,362.6 | 491,198.3 |  66,019 | 2,028,881.5 | 2,063,787 | 2,093,939.5 | 2,7


|                       |   `count` |      `mean` |     `std` |   `min` |       `25%` |     `50%` |       `75%` |     `max` |
|:----------------------|----------:|------------:|----------:|--------:|------------:|----------:|------------:|----------:|
| **mirrorNEQ_hits**    |        35 |     8,335.3 |   2,284.2 |     274 |     8,935.5 |     9,025 |     9,201.5 |     9,637 |
| **NEGmirror_hits**    |        35 |     8,335.3 |   2,365.4 |     291 |     8,520.5 |     8,640 |     8,788.5 |    14,761 |
| **POSmirror_hits**    |        35 |    39,682.8 |  10,892.7 |   1,340 |    42,532.5 |    43,095 |    43,863.0 |    45,422 |
| **superNEQ_hits**     |        35 |    90,676.6 |  22,644.0 |   3,065 |    93,545.0 |    95,298 |    96,641.0 |   126,486 |
| **RBdirect_hits**     |        35 |    90,676.6 |  23,052.2 |   2,970 |    92,957.0 |    94,501 |    95,765.0 |   136,138 |
| **not-RBdirect_hits** |        35 | 1,965,362.6 | 491,198.3 |  66,019 | 2,028,881.5 | 2,063,787 | 2,093,939.5 | 2,746,650 |



In [72]:
subtotals.T.describe().round(1).T.sort_values('min', ascending=False)

count      mean        std     min      25%      50%       75%  \
part                                                                    
Pcc19    6.0  399892.2   853038.0  9181.0  18261.0  71331.0   98611.5   
Pcc06    6.0  403423.3   860798.4  9061.0  18583.2  71888.5   99185.2   
Pcc08    6.0  399854.3   853501.1  8958.0  18349.5  70889.5   98264.0   
Pcc12    6.0  393786.2   839797.2  8921.0  17882.8  70583.0   97302.5   
Pcc22    6.0  388975.2   829426.1  8840.0  17882.0  69860.0   95894.8   
Pcc20    6.0  392922.5   838351.2  8818.0  18005.2  70190.0   96601.0   
Pcc05    6.0  389198.2   830324.4  8813.0  17659.8  69409.0   96049.8   
Pcc18    6.0  395498.7   844311.7  8764.0  17892.5  70336.5   97152.2   
Pcc23    6.0  391459.0   835080.6  8745.0  17941.5  69853.0   96575.0   
Pcc03    6.0  387480.2   826885.4  8739.0  17503.0  69126.5   95356.5   
Pcc24    6.0  388408.7   828166.3  8714.0  17873.5  69638.5   96025.8   
Pcc16    6.0  391320.7   834904.6  8685.0  18006.5  69851.0   96299.8   
Pcc04    6.0  386970.2   826166.2  8679.0  17557.5  68690.5   94955.0   
Nyt2     6.0  512287.0  1095989.6  8679.0  21322.8  83747.0  133725.0   
Pcc28    6.0  385567.0   822115.9  8675.0  17718.0  68958.5   95488.8   
Pcc07    6.0  383618.5   818178.8  8673.0  17546.2  68782.5   94545.0   
Pcc14    6.0  379564.7   809463.3  8640.0  17433.2  68104.0   93483.2   
Pcc26    6.0  376696.5   803474.2  8619.0  17161.5  67459.5   92822.2   
Pcc10    6.0  383343.2   817594.1  8597.0  17510.0  68674.0   94614.8   
Pcc02    6.0  379921.2   810487.1  8585.0  17442.0  67733.0   93653.5   
Pcc25    6.0  381354.3   813442.4  8582.0  17528.0  68255.5   93904.5   
Pcc09    6.0  385665.2   823014.3  8569.0  17518.0  68689.0   95019.2   
Pcc13    6.0  378632.3   807907.4  8562.0  17374.0  67515.5   93061.5   
Pcc11    6.0  386361.3   824641.2  8552.0  17452.5  68738.0   95173.8   
Pcc00    6.0  392385.5   837776.6  8541.0  17744.0  69559.0   96565.5   
Pcc29    6.0  377771.7   805451.6  8500.0  17352.0  67638.0   93559.5   
Pcc01    6.0  386547.3   824862.9  8498.0  17637.0  68783.0   95343.2   
Pcc21    6.0  380804.3   812764.2  8480.0  17429.8  67925.5   93372.0   
Pcc15    6.0  379024.7   808576.8  8447.0  17412.5  67679.0   93402.5   
Pcc27    6.0  376427.5   803127.8  8422.0  17138.2  67213.0   92759.0   
Pcc17    6.0  382815.3   816802.6  8410.0  17383.0  68320.0   94464.2   
Nyt1     6.0  361446.7   773324.5  5474.0  14797.8  58069.5   96104.8   
Apw      6.0  347117.3   751491.0  3805.0   8313.5  52049.5   89575.5   
PccVa    6.0   12357.5    26345.2   293.0    557.8   2202.5    3073.2   
PccTe    6.0   12338.8    26327.0   274.0    561.0   2170.5    3073.5   

             max  
part              
Pcc19  2139234.0  
Pcc06  2158603.0  
Pcc08  2140164.0  
Pcc12  2106105.0  
Pcc22  2080162.0  
Pcc20  2102322.0  
Pcc05  2082218.0  
Pcc18  2117053.0  
Pcc23  2094183.0  
Pcc03  2073494.0  
Pcc24  2077021.0  
Pcc16  2093696.0  
Pcc04  2071538.0  
Nyt2   2746650.0  
Pcc28  2061840.0  
Pcc07  2051873.0  
Pcc14  2030054.0  
Pcc26  2014969.0  
Pcc10  2050397.0  
Pcc02  2032504.0  
Pcc25  2039960.0  
Pcc09  2063787.0  
Pcc13  2025964.0  
Pcc11  2067797.0  
Pcc00  2100614.0  
Pcc29  2020064.0  
Pcc01  2068435.0  
Pcc21  2038043.0  
Pcc15  2027709.0  
Pcc27  2013996.0  
Pcc17  2048259.0  
Nyt1   1937910.0  
Apw    1878981.0  
PccVa    66074.0  
PccTe    66019.0

In [73]:
from source.utils.dataframes import get_mad, calculate_var_coeff, enhance_descrip, quartile_dispersion

In [74]:
enhance_descrip(subtotals).T

POSmirror_hits  mirrorNEQ_hits  not-RBdirect_hits  \
unique_forms            35.0            35.0               35.0   
mean                 39682.8          8335.3          1965362.6   
std                  10892.7          2284.2           491198.3   
min                   1340.0           274.0            66019.0   
25%                  42532.5          8935.5          2028881.5   
median               43095.0          9025.0          2063787.0   
75%                  43863.0          9201.5          2093939.5   
max                  45422.0          9637.0          2746650.0   
total              1388898.0        291735.0         68787692.0   
var_coeff                0.3             0.3                0.2   
range                44082.0          9363.0          2680631.0   
IQ_range              1330.5           266.0            65058.0   
Q_disper                 0.0             0.0                0.0   
MAD                    657.0           162.0            33733.0   
upper_fence          45858.8          9600.5          2191526.5   
lower_fence          40536.8          8536.5          1931294.5   

              superNEQ_hits  NEGmirror_hits  RBdirect_hits  
unique_forms           35.0            35.0           35.0  
mean                90676.6          8335.3        90676.6  
std                 22644.0          2365.4        23052.2  
min                  3065.0           291.0         2970.0  
25%                 93545.0          8520.5        92957.0  
median              95298.0          8640.0        94501.0  
75%                 96641.0          8788.5        95765.0  
max                126486.0         14761.0       136138.0  
total             3173681.0        291735.0      3173681.0  
var_coeff               0.2             0.3            0.3  
range              123421.0         14470.0       133168.0  
IQ_range             3096.0           268.0         2808.0  
Q_disper                0.0             0.0            0.0  
MAD                  1594.0           140.0         1456.0  
upper_fence        101285.0          9190.5        99977.0  
lower_fence         88901.0          8118.5        88745.0

In [77]:
cat_totals = subtotals.sum().sort_values().to_frame('Totals').convert_dtypes()
nb_show_table(cat_totals)
total_total = cat_totals.T.filter(like='direct').squeeze().sum()


|                       |   `Totals` |
|:----------------------|-----------:|
| **mirrorNEQ_hits**    |    291,735 |
| **NEGmirror_hits**    |    291,735 |
| **POSmirror_hits**    |  1,388,898 |
| **superNEQ_hits**     |  3,173,681 |
| **RBdirect_hits**     |  3,173,681 |
| **not-RBdirect_hits** | 68,787,692 |




|                       |   `Totals` |
|:----------------------|-----------:|
| **mirrorNEQ_hits**    |    291,735 |
| **NEGmirror_hits**    |    291,735 |
| **POSmirror_hits**    |  1,388,898 |
| **superNEQ_hits**     |  3,173,681 |
| **RBdirect_hits**     |  3,173,681 |
| **not-RBdirect_hits** | 68,787,692 |



In [78]:
subtotals.loc['Total', :] = subtotals.sum(axis=0)
subtotals.loc['Avg', :] = subtotals.mean(axis=0)
subtotals.filter(like='RBdirect').loc[:, 'Total'] = subtotals.sum(axis=1)
subtotals.loc[:, 'Avg'] = subtotals.mean(axis=1)
subtotals = subtotals.astype('int')
print(subtotals.to_markdown(intfmt=',', floatfmt=',.0f'))

| part   |   POSmirror_hits |   mirrorNEQ_hits |   not-RBdirect_hits |   superNEQ_hits |   NEGmirror_hits |   RBdirect_hits |        Avg |
|:-------|-----------------:|-----------------:|--------------------:|----------------:|-----------------:|----------------:|-----------:|
| Apw    |           17,294 |            3,805 |           1,878,981 |          86,805 |            5,320 |          90,499 |    347,117 |
| Nyt1   |           26,803 |            5,474 |           1,937,910 |          89,336 |           10,796 |          98,361 |    361,446 |
| Nyt2   |           41,008 |            8,679 |           2,746,650 |         126,486 |           14,761 |         136,138 |    512,287 |
| Pcc00  |           43,742 |            9,078 |           2,100,614 |          96,962 |            8,541 |          95,376 |    392,385 |
| Pcc01  |           43,095 |            9,151 |           2,068,435 |          95,634 |            8,498 |          94,471 |    386,547 |
| Pcc02  |           42,597

/tmp/ipykernel_13593/1056583261.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  subtotals.filter(like='RBdirect').loc[:, 'Total'] = subtotals.sum(axis=1)


In [79]:
# total_total = subtotals.filter(like='direct').sum().sum()
print(f'${total_total:,}$ hits remaining in superset polar comparison---_`ANYdirect` essentially_')

$71,961,373$ hits remaining in superset polar comparison---_`ANYdirect` essentially_


$71,961,373$ hits remaining in superset polar comparison---_`ANYdirect` essentially_


In [39]:
cat_totals['% of full superset'] = (cat_totals / total_total) * 100
cat_totals = cat_totals.sort_index().sort_values('Totals')
nb_show_table(cat_totals.convert_dtypes())


|                       |   `Totals` |   `% of full superset` |
|:----------------------|-----------:|-----------------------:|
| **NEGmirror_hits**    |    291,735 |                   0.20 |
| **mirrorNEQ_hits**    |    291,735 |                   0.20 |
| **POSmirror_hits**    |  1,388,898 |                   0.94 |
| **RBdirect_hits**     |  3,173,681 |                   2.15 |
| **superNEQ_hits**     |  3,173,681 |                   2.15 |
| **not-RBdirect_hits** | 68,787,692 |                  46.50 |




|                       |   `Totals` |   `% of full superset` |
|:----------------------|-----------:|-----------------------:|
| **NEGmirror_hits**    |    291,735 |                   0.20 |
| **mirrorNEQ_hits**    |    291,735 |                   0.20 |
| **POSmirror_hits**    |  1,388,898 |                   0.94 |
| **RBdirect_hits**     |  3,173,681 |                   2.15 |
| **superNEQ_hits**     |  3,173,681 |                   2.15 |
| **not-RBdirect_hits** | 68,787,692 |                  46.50 |



In [80]:
mir_totals = cat_totals.filter(like='mirror_hits', axis=0).Totals
mir_totals
cat_totals.loc[mir_totals.index, '% of mirror subset'] = mir_totals / mir_totals.sum() * 100
cat_totals.iloc[:, 1:] = cat_totals.iloc[:, 1:].round(1)
cat_totals

Totals  % of mirror subset
mirrorNEQ_hits       291735                <NA>
NEGmirror_hits       291735                17.4
POSmirror_hits      1388898                82.6
superNEQ_hits       3173681                <NA>
RBdirect_hits       3173681                <NA>
not-RBdirect_hits  68787692                <NA>

In [81]:
super_totals = cat_totals.filter(like='RBdirect', axis=0).Totals
print('## Size Discrepancies\n')
prepond = pd.DataFrame([
    [t.min() for t in (super_totals, mir_totals)],
    [t.max() for t in (super_totals, mir_totals)],
              ], columns=['superset', 'mirr_sub'], index=['neg_count', 'pos_count'], dtype='int'
             ).transpose()
prepond["N"] = prepond.sum(axis=1)

prepond['neg_proportion'] = (prepond.neg_count / prepond.N * 100).round(2).astype('string') + '%'
prepond['pos_proportion'] = (prepond.pos_count / prepond.N * 100).round(2).astype('string')+ '%'
prepond['pos_odds'] = '**' + (prepond.pos_count / prepond.neg_count).multiply(1).round(0).astype('int').astype('string') + ':1**'
# prepond['(neg_odds)'] = (prepond.neg_count / prepond.pos_count).multiply(1).round(2).astype('string') + ':1'

nb_show_table(prepond.T)
# preponderance_df.assign
# pd.Series(name='positive preponderance', data={f'Superset': f'~{int(round(super_totals.max() / super_totals.min(), 0))}:1 odds',
# f'Mirr Sub':  f'~{int(round(mir_totals.max() / mir_totals.min(), 0))}:1 odds'}).to_frame().assign(negated=[super_totals.min(), mir_totals.min()], 
#                                                                                                   positive=[super_totals.max(), mir_totals.max()])

## Size Discrepancies


|                    | `superset`   | `mirr_sub`   |
|:-------------------|:-------------|:-------------|
| **neg_count**      | 3173681      | 291735       |
| **pos_count**      | 68787692     | 1388898      |
| **N**              | 71961373     | 1680633      |
| **neg_proportion** | 4.41%        | 17.36%       |
| **pos_proportion** | 95.59%       | 82.64%       |
| **pos_odds**       | **22:1**     | **5:1**      |



## Size Discrepancies


| Polarity Approx.             |       $*E$ |      $@E$ |
|:-----------------------------|-----------:|----------:|
| **Negative Polarity Margin** |  3,173,681 |   291,735 |
| **Positive Polarity Margin** | 68,787,692 | 1,388,898 |
| $N$ (either)                 | 71,961,373 | 1,680,633 |
| **Negative Proportion**      |      4.41% |    17.36% |
| **Positive Proportion**      |     95.59% |    82.64% |
| **_Positive Preponderance_** |   **22:1** |   **5:1** |

<!-- 
| Polarity Approx. | Negative Polarity Margin | Positive Polarity Margin | `N` (either) |
|:-----------------|-------------------------:|-------------------------:|-------------:|
| $*E$             |                3,173,681 |               68,787,692 |   71,961,373 |
| $@E$             |                  291,735 |                1,388,898 |    1,680,633 |



|      | Negative Proportion | Positive Proportion | _Positive Preponderance_ |
|:----:|--------------------:|--------------------:|-------------------------:|
| $*E$ |               4.41% |              95.59% |                 **22:1** |
| $@E$ |              17.36% |              82.64% |                  **5:1** |
-->